ModelConfiguration

Atticus1806 opened the pull request:
This is a first implementation of the ModelConfiguration base class and some basic tests. 
Feel free to discuss, since we didn't have a strong opinion on the decisions most of the time.
E.g. we removed the `metadata` helpers since we decided they are most likely not used.
Also naming is up for discussion as well as documentation etc.

Marvin84 commented:
Just an additional comment that we discussed: the presented structure should have all we need for an atomic Config object. We can then construct nested configs for a better definition of the structure and hierarchy of/between different parameters.

michelwi commented on i6_models/assemblies/util.py:
```diff
@@ -0,0 +1,51 @@
+        return self.__dataclass_fields__[attribute_name].type
+
+    @classmethod
+    def from_namespace(cls, args):
```
do we really want to silently drop all config params that do not match? This seems error prone to me.

Why do we even need such a config class converter?

Atticus1806 replied:
So the usecase I found to not drop this was that you could have one big custom Config to use for your model which then initializes each subconfig you later need for the model.
So e.g. `TransformerConfig` can then initialize `TransformerEncoder` and `TransformerDecoder` etc. 
But if this makes sense or should be dropped is def. up for discussion.

curufinwe commented on i6_models/assemblies/util.py:
```diff
@@ -0,0 +1,51 @@
+from dataclasses import MISSING, dataclass
```
This should not live in assemblies. It can also be used in parts. I would put it in /i6_models/config.py

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+            return f.default_factory()
+        return f.default
+
+    def _get_type(self, attribute_name: str) -> Any:
```
Where is this function used?

Atticus1806 replied:
I don't understand. This really depends on your usecase. Here this is just abstract implementation for use in derived configs. Same goes for similar comments.

albertz replied:
I don't see any other code which uses this. So if some code is not used, I would remove it. We should not have code here which might be used later by other code but is not used yet.

Atticus1806 replied:
This class is supposed to offer basic functionality for configs and while you are right that we should carefully decide which functions are helpful and which are not, I think `_get_type` is something that will be used. Similar to the other 3 private functions. Regarding the name, maybe we should remove that for now, I am not sure.

albertz replied:
Ok, but then my question is, where will it be used, how will it be used? I don't really see any documentation here, nor any usage, so it's hard to just guess the potential usage. It would help to see some actual code where it is used, and maybe also some documentation.

As this is a private method, I assume the usage will only be internal. So what internal code would use it? Why did you not include this basic functionality?

Atticus1806 replied:
An example would be to check for proper initialization of the config. E.g. you would want `num_layers` to always be int. So instead of checking for each parameter individually you can check whether the value the user gives is the same type as the field.

albertz replied:
Who do you mean by "you"? The user? I thought this is only for internal use, i.e. the user would never use it. It the user is supposed to use it, it should not be marked as private, i.e. not start with underscore.

Is type checking supposed to be done anyway automatically? I would have expected that all types would always be checked, or not? But where is that code?

Atticus1806 replied:
>Is type checking supposed to be done anyway automatically?

No. This was just some example I can up with.

>I thought this is only for internal use

Yes. From this e.g. `ConformerConfig` is derived where then `_get_type` would be used (internally). Type checking was just an example I came up with.

Atticus1806 replied:
I think the major discussion point here is, which functions this base class should provide and which are deemed not helpful for now. Maybe we can discuss this in todays meeting, or other can comment? 
I don't really have a strong opinion on those function, those where just the ones I thought would be helpful looking at the Fairseq class.

albertz replied:
So, using the example:
```python
@dataclass
class ExampleConfig(ModelConfiguration):
    hidden_dim: int = 256
    name: str = "Example Configuration"

class ExampleModule(Module):
    def __init__(self, cfg: ExampleConfig)
        self.hidden_dim = cfg.hidden_dim
```
Now when the user writes:
```python
mod = ExampleModule(ExampleConfig(hidden_dim=2.5))
```
I thought that some code here would check the types and raise an exception here that `hidden_dim` is wrong. But this would be handled by `ModelConfiguration` automatically for all attribs. This is why I thought we have this `_get_type` method. Otherwise I still don't really understand why we need the `_get_type` method.

Or do you think that in `ExampleModule`, there would be an extra check like:
```python
assert isinstance(cfg.hidden_dim, cfg._get_type("hidden_dim"))
```
But this would be bad, as you would access a private method from outside. Also, this is not really nice to read. In this example, I would simply prefer:
```python
assert isinstance(cfg.hidden_dim, int)
```
Even if you would replicate the type `int` both here and in the `ExampleConfig` definition.

Or is this not the purpose of the function? I still don't really see the purpose. The only purpose which makes sense for me is if this would be used by some automatic type checking for all attribs. But that's why I was asking for some examples of potential usages.

SimBe195 replied:
> do you mean we should add code to the config class to do type checking automatically?

I think it's a good idea to force correct typing in the config and throw an exception if this is violated.

albertz replied:
I was also wondering whether `dataclass` itself already supports such a functionality. Maybe by defining `_get_type`, this is actually the way how `dataclass` would handle it? I don't know. That's why I was asking here.

albertz replied:
[Relevant StackOverflow question](https://stackoverflow.com/questions/50563546/validating-detailed-types-in-python-dataclasses).

michelwi replied:
a quick search brought me to this library that enables type checking at runtime with a decorator 
https://pypi.org/project/typeguard/

we could just add this to the dataclass and all should be checked automatically

albertz replied:
> a quick search brought me to this library that enables type checking at runtime with a decorator https://pypi.org/project/typeguard/
> 
> we could just add this to the dataclass and all should be checked automatically

Are you sure this works here?

And this would not need the `_get_type` method here, right?

Atticus1806 replied:
I am currently trying to set `typeguard` up, but I think it is not that straight forward.

Atticus1806 replied:
ESPnet actually wraps the `check_type` function. 
https://espnet.github.io/espnet/_modules/espnet2/utils/build_dataclass.html

albertz replied:
See my StackOverflow link above. It is specifically about possible solutions for `dataclass`.

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+    def _get_name(self, attribute_name: str) -> str:
+        return self.__dataclass_fields__[attribute_name].name
+
+    def _get_default(self, attribute_name: str) -> Any:
```
Where is this function used?

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+from dataclasses import MISSING, dataclass
```
I would add some docstring explaining (high-level) what the logic of the config would be.

Atticus1806 replied:
Do you want an example (e.g. one of the classes from the tests)? Because for now this is not really set I think. For now we wanted a base definition model configurations are built upon.

albertz replied:
But you must have some basic understanding of how this model config would be used, or not?

albertz replied:
It would also be helpful to document the inspiration from Fairseq.

Atticus1806 replied:
> But you must have some basic understanding of how this model config would be used

Yes, construction for example as in the tests. But for example then the later use for models and integration into our pipeline is not fully decided from my understanding. What I was asking was, whether you want an example or a rather conceptual explaining here.

albertz replied:
If possible, both, a conceptual explaining, and a minimal example.
I'm fine with the most minimal (1 line of code) example you can think of.
The conceptual explanation can also be short for now (one sentence).
A reference to the Fairseq code would also be very helpful.

albertz replied:
I also mean a costring for the module itself.

Atticus1806 replied:
So how the config then later is used in the module? This is highly up to the user and also not set yet. 
The idea discussed in the meeting was to provide a base config class similar to Fairseq, which can then be used to build the first modules. But this can be added later once we have a "common" way of using these configs.

albertz replied:
I'm fine with whatever minimal example you can think of, and whatever most minimal explanation you can give.

albertz replied:
Looks good!

Atticus1806 commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+
+    @staticmethod
+    def name() -> Optional[str]:
+        return None
```
If others agree this can def. be changed. For now this is just similar to the Fairseq reference implementation. But I assume that in this case yes each config should overwrite this if use is desired.

Atticus1806 replied:
For most of the functions here (and the ones I dropped) I found that they are barely used, but still I thought they provide nice optional functionality.

albertz replied:
But still, I don't understand the purpose of this function. It just returns `None`. So is this supposed to be overwritten? I don't understand the purpose. It this just for debugging purpose, or does this have any actual semantic meaning?

Atticus1806 replied:
So this is what I understand. But then I was not able to find a Config which overwrites this, so maybe we can remove it for now.

albertz replied:
Yes, I would remove code which we don't need for now. Once we need it, we can add it.

JackTemaki commented:
@Atticus1806 can you edit the initial post to include a link to the fairseq reference?

vieting commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+
+@dataclass
+class ModelConfiguration:
+    """base dataclass that supported fetching attributes"""
```
```suggestion
    """base dataclass that supports fetching attributes"""
```

vieting commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+        return self.__dataclass_fields__[attribute_name].type
+
+    @classmethod
+    def from_namespace(cls, args):
```
Can you please add type annotation?

Atticus1806 replied:
I am not 100% sure about the typing myself. So from what I understand this can be any object on which has attributes, so maybe `Any` is correct here?

vieting replied:
Yes, `Any` should be OK.

vieting commented on i6_models/config.py:
```diff
@@ -0,0 +1,51 @@
+    def from_namespace(cls, args):
+        """
+        Generates a ModelConfiguration from a given dataclass.
+        If its already of the correct instance it will return the input object, otherwise it creates an object
```
```suggestion
        If it's already of the correct instance, it will return the input object, otherwise it creates an object
```

SimBe195 commented on i6_models/config.py:
```diff
@@ -0,0 +1,63 @@
+Approach is inspired by the Fairseq: https://github.com/facebookresearch/fairseq/blob/main/fairseq/dataclass/configs.py
+"""
+
+from dataclasses import MISSING, dataclass
```
```suggestion
from dataclasses import MISSING, dataclass, fields
import typeguard
```

Atticus1806 commented:
I just added typechecks and a test which does that. Additionally I removed the private functions, since the opinion seems to be that we only add them when we need them.

Atticus1806 commented on i6_models/config.py:
```diff
@@ -0,0 +1,63 @@
+
+    def _get_type(self, attribute_name: str) -> Any:
+        return self.__dataclass_fields__[attribute_name].type
+
```
The difference between the two is that this way its always enforced, while with the decorator one could not apply it to certain configs and have those not type checked. I guess enforcing it is preferred?

SimBe195 commented on tests/test_config.py:
```diff
@@ -0,0 +1,85 @@
+    from typeguard import TypeCheckError
+
+    TestConfiguration(num_layers=2, hidden_dim=1)
+    try:
+        TestConfiguration(num_layers=2.0, hidden_dim="One")
+        assert False, "Typing should not allow this"
+    except TypeCheckError:
+        print("Typing Test worked")
```
```suggestion
    with pytest.raises(TypeCheckError):
        TestConfiguration(num_layers=2.0, hidden_dim="One")
```

vieting commented on tests/test_config.py:
```diff
@@ -0,0 +1,83 @@
+
+    from typeguard import TypeCheckError
+
+    TestConfiguration(num_layers=2, hidden_dim=1)
```
```suggestion
```
This isn't needed, is it?

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,56 @@
+        for field in fields(type(self)):
+            try:
+                typeguard.check_type(getattr(self, field.name), field.type)
+            except typeguard.TypeCheckError as e:
+                raise typeguard.TypeCheckError(f'In field "{field.name}" of "{type(self)}": {e}')
```
`e` is considered an ambiguous var name. I would use `exc`.
```suggestion
            except typeguard.TypeCheckError as exc:
                raise typeguard.TypeCheckError(f'In field "{field.name}" of "{type(self)}": {exc}')
```

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    @classmethod
+    def from_namespace(cls, args: Any):
```
Sorry to ask a bit late on this, but now I wonder: What is this function `from_namespace` actually used for?

Atticus1806 replied:
I would use it to initialize some configs with other existing ones. If I have a larger config with more parameters and want to initialize e.g. my encoder config with it, it can be done this way without having having to manually set all parameters again.

albertz replied:
Is this really a good idea? What is the benefit of doing this? I assume that you use this larger config for multiple different modules? But then I can imagine that there can be very nasty bugs or problems. E.g. when some option is maybe interpret differently depending on module (`dropout`, `momentum` and some others can be interpret as 1 - x or x), or some option name is not really consistent?

Didn't we actually discourage to do this for RASR, i.e. setting `*.momentum = 0.1` or so?

Can you add a docstring explaining this?

Atticus1806 replied:
I think also @michelwi wasn't a big fan of this either. This is what I would imagine to use it for, but if people rather prefer not to have this, due to the high chance of errors, then I can just remove it (and possible re-add it if we want this later on).
But then I would remove the Fairseq reference, because that was the last Fairseq part :D

albertz replied:
It's a bit difficult to discuss on this without seeing actual examples. Maybe you can find some in Fairseq?

@mmz33 do you know whether Zoltan uses this?

It still make sense to keep the Fairseq reference, but remove the "inspired from" part, just list it as a possible alternative, or extension. It might also make sense to list some other alternatives.

In general, it would be a good idea to have a short comparison to some of the popular alternatives. I know Lingvo also has such a configuration system (@mmz33 can maybe say sth about it). But I think there are also a few popular standalone packages for managing the configuration.

Btw, sorry if this question is a bit late: Did we actually look at some of these popular standalone packages, and did we consider to use just something existing, instead of doing our own thing here? Why did we decide against reusing sth existing?

Ok, when our `ModelConfiguration` is actually really minimal, like it is now, esp after removing `from_namespace`, it is not really needed.

Atticus1806 replied:
The Fairseq Transformer makes use of it but I feel its hart to follow the code because it not really documented, e.g:
https://github.com/facebookresearch/fairseq/blob/3f6ba43f07a6e9e2acf957fc24e57251a7a3f55c/fairseq/models/transformer/transformer_decoder.py#L467

In the first meeting we talked about this a bit and decided that we wanted to stay close to the style Fairseq does it.

albertz replied:
I rather meant examples of configurations where options are just put into a flat namespace. To see whether such configuration is easy to understand, or whether it rather makes things ambiguous.

michelwi replied:
> Btw, sorry if this question is a bit late: Did we actually look at some of these popular standalone packages, and did we consider to use just something existing, instead of doing our own thing here? Why did we decide against reusing sth existing?

We did talk about it. We decided to base the config on the fairseq config system.

> I think also @michelwi wasn't a big fan of this either. 

Yes i am very much afraid that this leads to silent bugs where a user thinks they have sweet a config option but in fact it has been dropped or misinterpreted.

Atticus1806 replied:
I can see that. Also I could not really find an easy to read example, but then I never worked with Fairseq and so maybe I just dont know where to look. But I will remove it for now then.

albertz replied:
> We decided to base the config on the fairseq config system.

Ok, basing it on this, but not really using it directly. I really mean to use sth existing, not maintaining our own code for this.

But I also mean other independent packages which are just for this purpose.

What alternatives did you consider? It would be good to list alternatives also in the docstring, to understand later why some decision was made.

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    @classmethod
+    def from_namespace(cls, args: Any):
```
```suggestion
    def from_namespace(cls, args: Any) -> ModelConfiguration:
```

albertz commented on i6_models/config.py:
```diff
@@ -0,0 +1,56 @@
+
+This config can then be used in the construction of the model to provide parameters.
+Approach is inspired by the Fairseq: https://github.com/facebookresearch/fairseq/blob/main/fairseq/dataclass/configs.py
+"""
```
Add empty line. And also import future annotation support.
```suggestion
"""

from __future__ import annotations
```
